File: Sources/App/Classes/Controllers/TXAppearance.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2018 - 2020 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 *	* Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual and/or Codeux Software, nor the names of
 *    its contributors may be used to endorse or promote products derived
 * 	  from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "NSObjectHelperPrivate.h"
#import "TPCPreferencesLocal.h"
#import "TXAppearance.h"

NS_ASSUME_NONNULL_BEGIN

NSString * const TXApplicationAppearanceChangedNotification = @"TXApplicationAppearanceChangedNotification";
NSString * const TXSystemAppearanceChangedNotification = @"TXSystemAppearanceChangedNotification";

@interface TXAppearancePropertyCollection ()
@property (nonatomic, copy, readwrite) NSString *appearanceName;
@property (nonatomic, assign, readwrite) TXAppearanceType appearanceType;
@property (nonatomic, assign, readwrite) BOOL isDarkAppearance;
@property (nonatomic, assign, readwrite) TXAppKitAppearanceTarget appKitAppearanceTarget;
@end

@interface TXAppearance ()
@property (nonatomic, strong, readwrite) TXAppearancePropertyCollection *properties;
@end

@implementation TXAppearance

#pragma mark -
#pragma mark Initialization

- (instancetype)init
{
	if ((self = [super init])) {
		[self prepareInitialState];

		return self;
	}

	return nil;
}

- (void)prepareInitialState
{
	[self updateAppearance];

	[RZWorkspaceNotificationCenter() addObserver:self
										selector:@selector(accessibilityDisplayOptionsDidChange:)
											name:NSWorkspaceAccessibilityDisplayOptionsDidChangeNotification
										  object:nil];

	[NSApp addObserver:self forKeyPath:@"effectiveAppearance" options:NSKeyValueObservingOptionNew context:NULL];
}

- (void)prepareForApplicationTermination
{
	LogToConsoleTerminationProgress("Removing appearance change observers");

	[RZNotificationCenter() removeObserver:self];

	[NSApp removeObserver:self forKeyPath:@"effectiveAppearance"];
}

#pragma mark -
#pragma mark Properties

+ (nullable NSString *)appearanceNameForType:(TXAppearanceType)type
{
	switch (type) {
		case TXAppearanceTypeLight:
		{
			return @"TahoeLight";
		}
		case TXAppearanceTypeDark:
		{
			return @"TahoeDark";
		}
		default:
		{
			return nil;
		}
	}
}

#pragma mark -
#pragma mark Notifications

- (void)observeValueForKeyPath:(nullable NSString *)keyPath ofObject:(nullable id)object change:(nullable NSDictionary<NSString *, id> *)change context:(nullable void *)context
{
	if ([keyPath isEqualToString:@"effectiveAppearance"]) {
		[self applicationAppearanceChanged];
	}
}

- (void)applicationAppearanceChanged
{
	/* Wait until next pass of the run loop to perform
	 update because the effective appearance may not
	 be propagated to all subviews when this is called. */
	XRPerformBlockAsynchronouslyOnMainQueue(^{
		[self updateAppearanceBySystemChange];
	});
}

- (void)systemColorsDidChange:(NSNotification *)aNote
{
	[self updateAppearanceBySystemChange];
}

- (void)accessibilityDisplayOptionsDidChange:(NSNotification *)aNote
{
	[self updateAppearanceBySystemChange];
}

- (void)updateAppearance
{
	[self updateAppearanceBySystemChange:NO];
}

- (void)updateAppearanceBySystemChange
{
	[self updateAppearanceBySystemChange:YES];
}

- (void)updateAppearanceBySystemChange:(BOOL)systemChanged
{
	TXAppearanceType appearanceType = TXAppearanceTypeLight;

	TXPreferredAppearance preferredAppearance = [TPCPreferences appearance];

	/* Determine user's preference */
	switch (preferredAppearance) {
		case TXPreferredAppearanceInherited:
		{
			if ([TXAppearancePropertyCollection systemWideDarkModeEnabled]) {
				appearanceType = TXAppearanceTypeDark;
			}

			break;
		}
		case TXPreferredAppearanceDark:
		{
			appearanceType = TXAppearanceTypeDark;

			break;
		}
		default:
		{
			break;
		}
	}

	BOOL isAppearanceDark = (appearanceType == TXAppearanceTypeDark);

	/* Determine best appearance inheritance approach */
	/* Before Mojave, appearance needs to be applied to every view.
	 After Mojave, views properly inherit the appearance of the parent
	 window or the system. If the user selects a specific appearance,
	 then we apply that to the parent window on Mojave and later to
	 allow views to inherit that. If the user selects the system
	 appearance, then we apply nothing and allow it to be inherited
	 from the system. */
	TXAppKitAppearanceTarget appKitAppearanceTarget = TXAppKitAppearanceTargetNone;

	if (preferredAppearance != TXPreferredAppearanceInherited) {
		appKitAppearanceTarget = TXAppKitAppearanceTargetWindow;
	}

	/* Test for changes */
	TXAppearancePropertyCollection *oldProperties = self.properties;

	BOOL changeAppearance = (oldProperties == nil ||
							 (oldProperties.appearanceType != appearanceType) ||
							 (oldProperties.appKitAppearanceTarget != appKitAppearanceTarget));

	if (changeAppearance == NO)
	{
		/* Even if the desired appearance hasn't changed, we still
		 signal views to perform selection update so that vibrant
		 views can draw correctly when the system changes. */

		if (systemChanged == NO) {
			return;
		}
	}
	else
	{
		/* When appearance changes as a result of a system change, then we
		 treat it as an application change as that's more specialized. */
		systemChanged = NO;
	}

	/* Assign new properties */
	TXAppearancePropertyCollection *newProperties = [TXAppearancePropertyCollection new];

	newProperties.appearanceName = [self.class appearanceNameForType:appearanceType];

	newProperties.appearanceType = appearanceType;

	newProperties.isDarkAppearance = isAppearanceDark;
	
	newProperties.appKitAppearanceTarget = appKitAppearanceTarget;

	self.properties = newProperties;

	if (preferredAppearance == TXPreferredAppearanceInherited) {
		NSApp.appearance = nil;
	} else if (isAppearanceDark) {
		NSApp.appearance = [TXAppearancePropertyCollection appKitDarkAppearance];
	} else {
		NSApp.appearance = [TXAppearancePropertyCollection appKitLightAppearance];
	}

	/* Notify observers */
	if (systemChanged == NO) {
		[self notifyApplicationAppearanceChanged];
	} else {
		[self notifySystemAppearanceChanged];
	}
}

- (void)notifyApplicationAppearanceChanged
{
	[RZNotificationCenter() postNotificationName:TXApplicationAppearanceChangedNotification object:self];
}

- (void)notifySystemAppearanceChanged
{
	[RZNotificationCenter() postNotificationName:TXSystemAppearanceChangedNotification object:self];
}

@end

#pragma mark -
#pragma mark Property Collection

@implementation TXAppearancePropertyCollection

- (nullable NSAppearance *)appKitAppearance
{
	if (self.appKitAppearanceTarget == TXAppKitAppearanceTargetNone) {
		return nil;
	}

	if (self.isDarkAppearance) {
		return [self.class appKitDarkAppearance];
	} else {
		return [self.class appKitLightAppearance];
	}
}

- (NSString *)shortAppearanceDescription
{
	if (self.isDarkAppearance == NO) {
		return @"light";
	} else {
		return @"dark";
	}
}

+ (BOOL)systemWideDarkModeEnabled
{
	return ([[NSApp effectiveAppearance] bestMatchFromAppearancesWithNames:@[NSAppearanceNameDarkAqua]] != nil);
}

+ (nullable NSAppearance *)appKitDarkAppearance
{
	return [NSAppearance appearanceNamed:NSAppearanceNameDarkAqua];
}

+ (nullable NSAppearance *)appKitLightAppearance
{
	return [NSAppearance appearanceNamed:NSAppearanceNameAqua];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/ApplicationSupport/TXAppearance.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2018 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit
import os

private let appearanceDefaultName = "Tahoe"

private let appearanceTerminationLogger = Logger(
	subsystem: Bundle.main.bundleIdentifier ?? "Glasstual",
	category: "Termination"
)

public extension Notification.Name {
	static let applicationAppearanceChanged = Notification.Name("TXApplicationAppearanceChangedNotification")
	static let systemAppearanceChanged = Notification.Name("TXSystemAppearanceChangedNotification")
}

@objc(TXAppearancePropertyCollection)
public final class AppearancePropertyCollection: NSObject, TXAppearanceProperties {
	@objc public var appearanceName = ""
	@objc public var appearanceType: TXAppearanceType = .light
	@objc public var isDarkAppearance = false
	@objc public var appKitAppearanceTarget: TXAppKitAppearanceTarget = .none

	@objc public var appKitAppearance: NSAppearance? {
		if appKitAppearanceTarget == .none {
			return nil
		}

		if isDarkAppearance {
			return Self.appKitDarkAppearance()
		}

		return Self.appKitLightAppearance()
	}

	@objc public var shortAppearanceDescription: String {
		isDarkAppearance ? "dark" : "light"
	}

	@MainActor @objc public static func systemWideDarkModeEnabled() -> Bool {
		NSApp.effectiveAppearance.bestMatch(from: [NSAppearance.Name.darkAqua]) != nil
	}

	@objc public static func appKitDarkAppearance() -> NSAppearance? {
		NSAppearance(named: .darkAqua)
	}

	@objc public static func appKitLightAppearance() -> NSAppearance? {
		NSAppearance(named: .aqua)
	}
}

@objc(TXAppearance)
@MainActor
public final class Appearance: NSObject {
	@objc public private(set) var properties: AppearancePropertyCollection!

	private var isApplyingAppearance = false
	private var effectiveAppearanceObservation: NSKeyValueObservation?

	override public init() {
		super.init()
		prepareInitialState()
	}

	deinit {
		MainActor.assumeIsolated {
			removeObservers()
		}
	}

	private func prepareInitialState() {
		updateAppearance()

		NSWorkspace.shared.notificationCenter.addObserver(
			self,
			selector: #selector(accessibilityDisplayOptionsDidChange(_:)),
			name: NSWorkspace.accessibilityDisplayOptionsDidChangeNotification,
			object: nil
		)

		effectiveAppearanceObservation = NSApp.observe(\.effectiveAppearance, options: .new) { [weak self] _, _ in
			MainActor.assumeIsolated {
				guard let self, self.isApplyingAppearance == false else {
					return
				}
				self.applicationAppearanceChanged()
			}
		}
	}

	@objc public func prepareForApplicationTermination() {
		appearanceTerminationLogger.debug("Removing appearance change observers")
		removeObservers()
	}

	private func removeObservers() {
		guard let effectiveAppearanceObservation else {
			return
		}

		self.effectiveAppearanceObservation = nil
		NSWorkspace.shared.notificationCenter.removeObserver(self)
		effectiveAppearanceObservation.invalidate()
	}

	private func applicationAppearanceChanged() {
		Task { @MainActor [weak self] in
			self?.updateAppearanceBySystemChange(true)
		}
	}

	@objc private func accessibilityDisplayOptionsDidChange(_: Notification) {
		updateAppearanceBySystemChange(true)
	}

	@objc public func updateAppearance() {
		updateAppearanceBySystemChange(false)
	}

	private func updateAppearanceBySystemChange(_ systemChanged: Bool) {
		var appearanceType: TXAppearanceType = .light
		let preferredAppearance = TextualPreferences.appearance()

		switch preferredAppearance {
		case .inherited:
			applyAppKitAppearance(nil)

			if AppearancePropertyCollection.systemWideDarkModeEnabled() {
				appearanceType = .dark
			}
		case .dark:
			appearanceType = .dark
		default:
			break
		}

		let isAppearanceDark = appearanceType == .dark

		var appKitAppearanceTarget: TXAppKitAppearanceTarget = .none
		if preferredAppearance != .inherited {
			appKitAppearanceTarget = .window
		}

		let oldProperties = properties
		let changeAppearance =
			oldProperties == nil
				|| oldProperties!.appearanceType != appearanceType
				|| oldProperties!.appKitAppearanceTarget != appKitAppearanceTarget

		var systemChanged = systemChanged

		if changeAppearance == false {
			if systemChanged == false {
				return
			}
		} else {
			systemChanged = false
		}

		let newProperties = AppearancePropertyCollection()
		newProperties.appearanceName = appearanceDefaultName
		newProperties.appearanceType = appearanceType
		newProperties.isDarkAppearance = isAppearanceDark
		newProperties.appKitAppearanceTarget = appKitAppearanceTarget
		properties = newProperties

		if preferredAppearance == .inherited {
			applyAppKitAppearance(nil)
		} else if isAppearanceDark {
			applyAppKitAppearance(AppearancePropertyCollection.appKitDarkAppearance())
		} else {
			applyAppKitAppearance(AppearancePropertyCollection.appKitLightAppearance())
		}

		if systemChanged {
			notifySystemAppearanceChanged()
		} else {
			notifyApplicationAppearanceChanged()
		}
	}

	private func applyAppKitAppearance(_ appearance: NSAppearance?) {
		let current = NSApp.appearance

		if current === appearance {
			return
		}

		if let current, let appearance, current.name == appearance.name {
			return
		}

		isApplyingAppearance = true
		NSApp.appearance = appearance
		isApplyingAppearance = false
	}

	private func notifyApplicationAppearanceChanged() {
		NotificationCenter.default.post(
			name: .applicationAppearanceChanged,
			object: self
		)
	}

	private func notifySystemAppearanceChanged() {
		NotificationCenter.default.post(
			name: .systemAppearanceChanged,
			object: self
		)
	}
}
